### PR TITLE
ROE-2388 Presenter back link

### PR DESCRIPTION
**Pull request by mwestacott**
### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/overseas.name.controller.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
       } else {
         await updateOverseasEntity(req, session);
       }
-      if (isActiveFeature(config.FEATURE_FLAG_ENABLE_REDIS_REMOVAL)){
+      if (isActiveFeature(config.FEATURE_FLAG_ENABLE_REDIS_REMOVAL)) {
```
Comment (mwestacott) [inline src/controllers/overseas.name.controller.ts:54]: Just fixes some formatting.

File: src/controllers/presenter.controller.ts [hunk 2]
```diff
@@ -6,7 +6,12 @@ import { isActiveFeature } from "../utils/feature.flag";
 import { getUrlWithParamsToPath } from "../utils/url";
 
 export const get = (req: Request, res: Response, next: NextFunction) => {
-  getPresenterPage(req, res, next, config.PRESENTER_PAGE, config.OVERSEAS_NAME_URL);
+  let backLinkUrl = config.OVERSEAS_NAME_URL;
+  if (isActiveFeature(config.FEATURE_FLAG_ENABLE_REDIS_REMOVAL)){
+    backLinkUrl = getUrlWithParamsToPath(config.OVERSEAS_NAME_WITH_PARAMS_URL, req);
```
Comment (mwestacott) [inline src/controllers/presenter.controller.ts:11]: This ensures back link is correctly rendered when page initially displays.

File: src/middleware/validation.middleware.ts [hunk 3]
```diff
@@ -56,7 +56,7 @@ export function checkValidations(req: Request, res: Response, next: NextFunction
       const entityNumber = appData?.[EntityNumberKey];
 
       return res.render(NAVIGATION[routePath].currentPage, {
-        backLinkUrl: NAVIGATION[routePath].previousPage(appData),
+        backLinkUrl: NAVIGATION[routePath].previousPage(appData, req),
```
Comment (mwestacott) [inline src/middleware/validation.middleware.ts:59]: Needed for when page redisplays due to validation errors.

File: src/utils/navigation.ts [hunk 4]
```diff
@@ -129,14 +130,19 @@ export const NAVIGATION: Navigation = {
...
+  },
   [config.PRESENTER_URL]: {
     currentPage: config.PRESENTER_PAGE,
     previousPage: () => config.OVERSEAS_NAME_URL,
     nextPage: [config.WHO_IS_MAKING_FILING_URL]
   },
   [config.PRESENTER_WITH_PARAMS_URL]: {
     currentPage: config.PRESENTER_PAGE,
-    previousPage: () => config.OVERSEAS_NAME_URL,
+    previousPage: (appData: ApplicationData, req: Request) => getUrlWithParamsToPath(config.OVERSEAS_NAME_WITH_PARAMS_URL, req),
```
Comment (mwestacott) [inline src/utils/navigation.ts:145]: This is why 'previousPage' function needed the request object - to build the correct link when back link is being rendered following a validation failure.

File: test/controllers/presenter.controller.spec.ts [hunk 5]
```diff
@@ -52,7 +54,7 @@ import { isActiveFeature } from "../../src/utils/feature.flag";
 import { serviceAvailabilityMiddleware } from "../../src/middleware/service.availability.middleware";
 import { getUrlWithParamsToPath } from "../../src/utils/url";
 
-const NEXT_PAGE_URL = "/NEXT_PAGE";
+const MOCKED_PAGE_URL = "/MOCKED_PAGE";
```
Comment (mwestacott) [inline test/controllers/presenter.controller.spec.ts:57]: Mocked page now representing a page being used when going forwards and backwards, so renamed.

File: test/controllers/presenter.controller.spec.ts [hunk 6]
```diff
@@ -255,32 +284,39 @@ describe("PRESENTER controller", () => {
 
   describe("POST with url params tests", () => {
     test(`redirect to the ${WHO_IS_MAKING_FILING_PAGE} page after a successful post from presenter page with url params`, async () => {
-      mockIsActiveFeature.mockReturnValueOnce(true); // For FEATURE_FLAG_ENABLE_ROE_UPDATE
+      mockIsActiveFeature.mockReturnValueOnce(true); // For FEATURE_FLAG_ENABLE_REDIS_REMOVAL
```
Comment (mwestacott) [inline test/controllers/presenter.controller.spec.ts:287]: This was always wrong - a copy/paste error when we mobbed it I think. Correcting.

**Review by HatiDev — APPROVED** [review]
looks good to me

**Review by dlloyd1-ch — APPROVED** [review]

**Pull request merged**